polygeist/lib/Dialect/Polygeist/Utils/TransformUtils.cpp

```diff
@@ -562,8 +562,17 @@ createSYCLAccessorSubscriptOp(TypedValue<MemRefType> accessor,
                               Location loc) {
   const auto accTy =
       cast<sycl::AccessorType>(accessor.getType().getElementType());
-  const auto MT = cast<MemRefType>(
-      cast<LLVM::LLVMStructType>(accTy.getBody()[1]).getBody()[0]);
+  assert(accTy.getDimension() != 0 && "Dimensions cannot be zero");
+  const auto ptrTy =
+      cast<LLVM::LLVMStructType>(accTy.getBody()[1]).getBody()[0];
+  unsigned memorySpace =
+      TypeSwitch<mlir::Type, unsigned>(ptrTy)
+          .Case<LLVM::LLVMPointerType>(
+              [](auto ty) { return ty.getAddressSpace(); })
+          .Case<MemRefType>([](auto ty) { return ty.getMemorySpaceAsInt(); });
```

Comment:
Also, this might be a better fit for a different patch

Reply:
Here we might be creating `memref<?xllvm.struct<...>>` types, which is not legal. Can we create `llvm.ptr` (opaque) for those cases? That might require changes to operations, AFAIK...

Reply:
Created https://github.com/intel/llvm/pull/9202.

Reply:
> Here we might be creating memref<?xllvm.struct<...>> types, which is not legal. Can we create llvm.ptr (opaque) for those cases? That might require changes to operations, AFAIK...

Right, we may generate `memref<?xllvm.struct<...>>` types if the buffer element type is a struct type. 
We are generating `memref<?xllvm.struct<...>> all over the places, e.g., the type argument promotion look for is exactly that type. I suggest we handle that generally. WDYT?

Reply:
> Instead of accessing the body, we might want to use the mapping in [-convert-sycl-to-llvm](https://github.com/intel/llvm/blob/ade4b093b3010dc3428312c3e8ed0e5c0497a7df/mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp#L95). That will also make our lives easier when we transition to the new attributes.

done.

Reply:
> Right, we may generate `memref<?xllvm.struct<...>>` types if the buffer element type is a struct type.
> We are generating `memref<?xllvm.struct<...>> all over the places, e.g., the type argument promotion look for is exactly that type. I suggest we handle that generally. WDYT?

I think this PR need not concern itself with this limitation. One way to work around the limitation imposed by the memref dialect is to create a polygeist type for a struct, add the required interface to allow that type to be the element type of a memref and then lower that type to llvm,struct while converting the polygeist dialect to the LLVM dialect.

Reply:
I agree with @etiotto. We should handle this in a different PR, for sure. Adding a `polygeist.struct` is the ideal solution IMO. We won't run into issues with this as long as we use `polygeist-opt` or `cgeist`, as we're programmatically allowing `memref<llvm.struct>` there. Note this is not standard, though, and we will need to address. LGTM